stdlib/public/core/CollectionAlgorithms.swift

```diff
@@ -260,6 +260,115 @@ extension MutableCollection where Self : BidirectionalCollection {
   }
 }
 
+//===----------------------------------------------------------------------===//
+// shuffled()/shuffle()
+//===----------------------------------------------------------------------===//
+
+extension Sequence {
+  /// Returns the elements of the sequence, shuffled using the given generator
+  /// as a source for randomness.
+  ///
+  /// You use this method to randomize the elements of a sequence when you
+  /// are using a custom random number generator. For example, you can shuffle
+  /// the numbers between `0` and `9` by calling the `shuffled(using:)` method
+  /// on that range:
+  ///
+  ///     let numbers = 0...9
+  ///     let shuffledNumbers = numbers.shuffled(using: &myGenerator)
+  ///     // shuffledNumbers == [8, 9, 4, 3, 2, 6, 7, 0, 5, 1]
+  ///
+  /// - Parameter generator: The random number generator to use when shuffling
+  ///   the sequence.
+  /// - Returns: An array of this sequence's elements in a shuffled order.
+  ///
+  /// - Complexity: O(*n*)
+  @inlinable
+  public func shuffled<T: RandomNumberGenerator>(
+    using generator: inout T
+  ) -> [Element] {
+    var result = ContiguousArray(self)
+    result.shuffle(using: &generator)
+    return Array(result)
+  }
+  
+  /// Returns the elements of the sequence, shuffled.
+  ///
+  /// For example, you can shuffle the numbers between `0` and `9` by calling
+  /// the `shuffled()` method on that range:
+  ///
+  ///     let numbers = 0...9
+  ///     let shuffledNumbers = numbers.shuffled()
+  ///     // shuffledNumbers == [1, 7, 6, 2, 8, 9, 4, 3, 5, 0]
+  ///
+  /// This method uses the default random generator, `Random.default`. The call
+  /// to `numbers.shuffled()` above is equivalent to calling
+  /// `numbers.shuffled(using: &Random.default)`.
+  ///
+  /// - Parameter generator: The random number generator to use when shuffling
+  ///   the sequence.
+  /// - Returns: A shuffled array of this sequence's elements.
+  ///
+  /// - Complexity: O(*n*)
+  @inlinable
+  public func shuffled() -> [Element] {
+    return shuffled(using: &Random.default)
+  }
+}
+
+extension MutableCollection {
+  /// Shuffles the collection in place, using the given generator as a source
+  /// for randomness.
+  ///
+  /// You use this method to randomize the elements of a collection when you
+  /// are using a custom random number generator. For example, you can use the
+  /// `shuffle(using:)` method to randomly reorder the elements of an array.
+  ///
+  ///     var names = ["Alejandro", "Camila", "Diego", "Luciana", "Luis", "Sofía"]
+  ///     names.shuffle(using: &myGenerator)
+  ///     // names == ["Sofía", "Alejandro", "Camila", "Luis", "Diego", "Luciana"]
+  ///
+  /// - Parameter generator: The random number generator to use when shuffling
+  ///   the collection.
+  ///
+  /// - Complexity: O(*n*)
```

Comment:
Hmmm. If the collection isn't random-access, this ends up being O(n^2), since `index(_:offsetBy:)` gets called *n* times. For random-access collections, it's O(n), since there are `count - 1` swaps.

Reply:
Accessing the `count` property can be O(*n*), so could you store it locally with name shadowing?

```swift
let count = self.count
guard count > 1 else { return }
// etc.
```

Reply:
Yes, that's an optimization that should be added at some point.

Reply:
or should I do something like this:
```swift
/// - Complexity: O(*n*) if the collection conforms to
///   `RandomAccessCollection`; otherwise, O(*n^2*), where *n* is the length
///   of the collection.
```

Reply:
To be honest, I was thinking we should probably give it the same constraints as `sort`, which is to add `RandomAccessCollection` as a requirement for the in-place shuffling methods.

@airspeedswift What do you think? Does parity with `sort` make sense?